models/user_repo_unit.go

```diff
@@ -0,0 +1,1079 @@
+
+// Except for the special users UserRepoUnitLoggedInUser and UserRepoUnitAnyUser,
+// only real users have records in the table (i.e. organizations don't have their
+// own records).
+
+// Users that are not active yet or have prohibited login do not have any records.
+
+// Restricted users are users that require explicit permissions to access a repository;
+// these users are solved by checking records that match their UserID but _ignoring_
+// records for UserRepoUnitLoggedInUser and UserRepoUnitAnyUser.
+
+// If all permissions result in AccessModeNone, the whole record is omitted,
+// so checking against the UserRepoUnit table will result in a quick check for
+// whether the user can see the repository at all (e.g. at the explore page).
+
+// Input considered for calculating the records includes:
+//
+// * public/private status of the repository
+// * public/limited/private status of the repository's owner (user or org)
+// * user team membership (if repository owner is an org)
+// * team settings (whether the team has an explicit repository list or they can access all of them)
+// * user collaboration status
+// * user ownership of the repository
+// * user site admin status
+// * which repository units are enabled (e.g. issues, PRs, etc.)
+
+// UserRepoUnit is an explosion (cartesian product) of all user permissions
+// on all repositories, and all types of relevant unit types.
+type UserRepoUnit struct {
+	UserID               int64      `xorm:"pk"`
+	RepoID               int64      `xorm:"pk INDEX"`
+	Type                 UnitType   `xorm:"pk"`
+	Mode                 AccessMode `xorm:"NOT NULL"`
+}
+
+// UserRepoUnitWork is a table used for temporarily accumulate all the work performed
+// while processing a batch. Ideally, this would be a temporary (no storage) table.
+// Records are grouped by BatchID in order to prevent any kind of collision.
+// Lack of primary key is intentional; this table is not intended for replication
+// as it should never contain rows after the transaction is completed.
+type UserRepoUnitWork struct {
+	BatchID              int64      `xorm:"NOT NULL INDEX"`
+	UserID               int64      `xorm:"NOT NULL"`
+	RepoID               int64      `xorm:"NOT NULL"`
+	Type                 UnitType   `xorm:"NOT NULL"`
+	Mode                 AccessMode `xorm:"NOT NULL"`
+}
+
+// UserRepoUnitBatchNumber provides in a safe way unique ID values
+// for the batch number in case we are in a multi-server environment.
+// It's a 63-bit number, so good luck reaching the maximum value
+// (300 million years at 1000 requests per second, if you want to know).
+// It's a makeshift replacement for an actual database sequence.
+type UserRepoUnitBatchNumber struct {
+	ID int64 `xorm:"pk autoincr"`
+}
+
+var (
+	// This list may be updated as needed. Requisites are:
+	// - New units need to abide by the same rules as the others
+	// - Units from all repos must be rebuilt after this change
+	supportedUnits = []UnitType {
```

Comment:
What is this list for? Why only these units?

Reply:
The list is for choosing what kinds of units we will actually be checking using this table (e.g. in listings, link generation, and such). There are some units that don't make much sense caching, like `UnitTypeExternalTracker`, perhaps? Or surely `UnitTypeReleases`, that can't be disabled in a user by user basis. Since I didn't get to the part where I'm actually _using_ this information, I still don't know which units makes sense to include, but I'd like to avoid processing more units than necessary.

Reply:
So with this change you would still keep the `Access` table?
If you included all units you can use it in `getUserRepoPermission`, you don't think it is worth it?

Reply:
> 
> 
> So with this change you would still keep the `Access` table?
> If you included all units you can use it in `getUserRepoPermission`, you don't think it is worth it?

Yes, this will end as a replacement of the `access` table. And yes, all units we expect to use with `getUserRepoPermission` should go in here. The thing is (AFAIU) that some units exists for repositories to have, not for users. From the back of my head I believe we can't deny `UnitTypeReleases` to a particular user or group, so that unit I'd leave out. But including one unit is only a matter of adding it to the list. If we were to include all units, we'd just leave `userRepoUnitSupported` empty.

I'm just trying to optimize storage and SQL operations, but if we think that the performance hit won't be that big of a deal, we can certainly enable them all. Those are easy changes.

Reply:
At least team setting page allow specifying access to Release unit.